File: core/src/protocol.rs
use crate::game::{PlayerId, RoomSnapshot};
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum ClientMessage {
    JoinOrCreateRoom {
        #[serde(rename = "playerName")]
        player_name: Option<String>,
        #[serde(rename = "roomCode")]
        room_code: Option<String>,
        #[serde(rename = "gameMode")]
        game_mode: Option<String>,
        #[serde(rename = "matchDurationSecs")]
        match_duration_secs: Option<u64>,
    },
    RejoinRoom {
        #[serde(rename = "rejoinToken")]
        rejoin_token: String,
    },
    InputUpdate {
        text: String,
    },
    SubmitAttempt {
        text: String,
    },
    StartMatch,
    Rematch,
}

#[derive(Debug, Serialize, Clone)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum ServerMessage {
    Welcome {
        #[serde(rename = "playerId")]
        player_id: PlayerId,
        #[serde(rename = "roomCode")]
        room_code: String,
        #[serde(rename = "gameKey")]
        game_key: String,
        #[serde(rename = "inputPlaceholder")]
        input_placeholder: String,
        #[serde(rename = "rejoinToken")]
        rejoin_token: String,
    },
    RoomState {
        room: RoomSnapshot,
    },
    PromptState {
        #[serde(rename = "roomCode")]
        room_code: String,
        #[serde(rename = "roundId")]
        round_id: u64,
        prompt: String,
    },
    RaceProgress {
        #[serde(rename = "roomCode")]
        room_code: String,
        #[serde(rename = "playerId")]
        player_id: PlayerId,
        text: String,
    },
    RoundResult {
        #[serde(rename = "roomCode")]
        room_code: String,
        #[serde(rename = "roundId")]
        round_id: u64,
        #[serde(rename = "winnerPlayerId")]
        winner_player_id: PlayerId,
        #[serde(rename = "growthAwarded")]
        growth_awarded: f32,
    },
    WrongAnswer {
        #[serde(rename = "roomCode")]
        room_code: String,
        #[serde(rename = "playerId")]
        player_id: PlayerId,
        #[serde(rename = "shrinkApplied")]
        shrink_applied: f32,
    },
    Error {
        message: String,
    },
}

#[cfg(test)]
mod tests {
    use super::ClientMessage;

    #[test]
    fn parses_all_supported_client_messages() {
        let join = r#"{"type":"joinOrCreateRoom","playerName":"Alice","roomCode":"ABCD","gameMode":"keyboarding","matchDurationSecs":90}"#;
        assert!(serde_json::from_str::<ClientMessage>(join).is_ok());

        let rejoin = r#"{"type":"rejoinRoom","rejoinToken":"abc123"}"#;
        assert!(serde_json::from_str::<ClientMessage>(rejoin).is_ok());

        let update = r#"{"type":"inputUpdate","text":"hel"}"#;
        assert!(serde_json::from_str::<ClientMessage>(update).is_ok());

        let submit = r#"{"type":"submitAttempt","text":"hello"}"#;
        assert!(serde_json::from_str::<ClientMessage>(submit).is_ok());

        let start = r#"{"type":"startMatch"}"#;
        assert!(serde_json::from_str::<ClientMessage>(start).is_ok());

        let rematch = r#"{"type":"rematch"}"#;
        assert!(serde_json::from_str::<ClientMessage>(rematch).is_ok());
    }

    #[test]
    fn rejects_removed_ping_message() {
        let ping = r#"{"type":"ping","sentAtMs":123}"#;
        assert!(serde_json::from_str::<ClientMessage>(ping).is_err());
    }

    #[test]
    fn serializes_wrong_answer_message() {
        let msg = super::ServerMessage::WrongAnswer {
            room_code: "ABCD".to_string(),
            player_id: 1,
            shrink_applied: 2.0,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"wrongAnswer""#));
        assert!(json.contains(r#""roomCode":"ABCD""#));
        assert!(json.contains(r#""playerId":1"#));
        assert!(json.contains(r#""shrinkApplied":2.0"#));
    }
}

File: client/src/lib/game/protocol.ts
export type PlayerSnapshot = {
	id: number;
	name: string;
	size: number;
	color: string;
	connected: boolean;
	progress: string;
};

export type RoomSnapshot = {
	roomCode: string;
	players: PlayerSnapshot[];
	prompt: string;
	roundId: number;
	matchWinner: number | null;
	matchRemainingMs: number | null;
	hostPlayerId: number;
};

export type ClientMessage =
	| {
			type: 'joinOrCreateRoom';
			playerName?: string;
			roomCode?: string;
			gameMode?: string;
			matchDurationSecs?: number;
	  }
	| { type: 'rejoinRoom'; rejoinToken: string }
	| { type: 'inputUpdate'; text: string }
	| { type: 'submitAttempt'; text: string }
	| { type: 'startMatch' }
	| { type: 'rematch' };

export type ServerMessage =
	| {
			type: 'welcome';
			playerId: number;
			roomCode: string;
			gameKey: string;
			inputPlaceholder: string;
			rejoinToken: string;
	  }
	| { type: 'roomState'; room: RoomSnapshot }
	| { type: 'promptState'; roomCode: string; roundId: number; prompt: string }
	| { type: 'raceProgress'; roomCode: string; playerId: number; text: string }
	| {
			type: 'roundResult';
			roomCode: string;
			roundId: number;
			winnerPlayerId: number;
			growthAwarded: number;
	  }
	| { type: 'wrongAnswer'; roomCode: string; playerId: number; shrinkApplied: number }
	| { type: 'error'; message: string };

function isObject(value: unknown): value is Record<string, unknown> {
	return typeof value === 'object' && value !== null;
}

function isPlayerSnapshot(value: unknown): value is PlayerSnapshot {
	if (!isObject(value)) return false;
	return (
		typeof value.id === 'number' &&
		typeof value.name === 'string' &&
		typeof value.size === 'number' &&
		typeof value.color === 'string' &&
		typeof value.connected === 'boolean' &&
		typeof value.progress === 'string'
	);
}

function isRoomSnapshot(value: unknown): value is RoomSnapshot {
	if (!isObject(value) || !Array.isArray(value.players)) return false;
	return (
		typeof value.roomCode === 'string' &&
		typeof value.prompt === 'string' &&
		typeof value.roundId === 'number' &&
		(value.matchWinner === null || typeof value.matchWinner === 'number') &&
		(value.matchRemainingMs === null || typeof value.matchRemainingMs === 'number') &&
		typeof value.hostPlayerId === 'number' &&
		value.players.every(isPlayerSnapshot)
	);
}

function isServerMessage(value: unknown): value is ServerMessage {
	if (!isObject(value) || typeof value.type !== 'string') {
		return false;
	}
	switch (value.type) {
		case 'welcome':
			return (
				typeof value.playerId === 'number' &&
				typeof value.roomCode === 'string' &&
				typeof value.gameKey === 'string' &&
				typeof value.inputPlaceholder === 'string' &&
				typeof value.rejoinToken === 'string'
			);
		case 'roomState':
			return isRoomSnapshot(value.room);
		case 'promptState':
			return (
				typeof value.roomCode === 'string' &&
				typeof value.roundId === 'number' &&
				typeof value.prompt === 'string'
			);
		case 'raceProgress':
			return (
				typeof value.roomCode === 'string' &&
				typeof value.playerId === 'number' &&
				typeof value.text === 'string'
			);
		case 'roundResult':
			return (
				typeof value.roomCode === 'string' &&
				typeof value.roundId === 'number' &&
				typeof value.winnerPlayerId === 'number' &&
				typeof value.growthAwarded === 'number'
			);
		case 'wrongAnswer':
			return (
				typeof value.roomCode === 'string' &&
				typeof value.playerId === 'number' &&
				typeof value.shrinkApplied === 'number'
			);
		case 'error':
			return typeof value.message === 'string';
		default:
			return false;
	}
}

export function decodeServerMessage(raw: string): ServerMessage | null {
	try {
		const parsed: unknown = JSON.parse(raw);
		return isServerMessage(parsed) ? parsed : null;
	} catch {
		return null;
	}
}
